openBCIBoard.js

```diff
@@ -856,180 +1013,174 @@ function OpenBCIFactory() {
+
+                // Over write data
+                data = newDataBuffer;
+
+                // Update the number of bytes to read
+                bytesToRead += oldBytesWritten;
+            }
+
+            var readingPosition = 0;
+
+            // 45 < (200 - 33) --> 45 < 167 (good) | 189 < 167 (bad) | 0 < (28 - 33) --> 0 < -5 (bad)
+            while (readingPosition <= bytesToRead - k.OBCIPacketSize) {
+                if (data[readingPosition] === k.OBCIByteStart) {
+                    var rawPacket = data.slice(readingPosition, readingPosition + k.OBCIPacketSize);
+                    if (data[readingPosition + k.OBCIPacketSize - 1] === k.OBCIByteStop) {
+                        // standard packet!
+                        openBCISample.parseRawPacket(rawPacket,this.channelSettingsArray)
+                            .then(sampleObject => {
+
+                                sampleObject._count = this.sampleCount++;
+                                if(this.impedanceTest.active) {
+                                    var impedanceArray;
+                                    if (this.impedanceTest.continuousMode) {
+                                        //console.log('running in contiuous mode...');
+                                        //openBCISample.debugPrettyPrint(sampleObject);
+                                        impedanceArray = openBCISample.goertzelProcessSample(sampleObject,this.goertzelObject)
+                                        if (impedanceArray) {
+                                            this.emit('impedanceArray',impedanceArray);
+                                        }
+                                    } else if (this.impedanceTest.onChannel != 0) {
+                                        // Only calculate impedance for one channel
+                                        impedanceArray = openBCISample.goertzelProcessSample(sampleObject,this.goertzelObject)
+                                        if (impedanceArray) {
+                                            this.impedanceTest.impedanceForChannel = impedanceArray[this.impedanceTest.onChannel - 1];
+                                        }
+                                    }
+                                } else {
+                                    this.emit('sample', sampleObject);
+                                }
                             });
-                        }
                     }
-                    this.emit('sample', newSample);
-                } else {
-                    this.badPackets++;
-                    this._bufAlign(); // fix the buffer to start reading at next start byte
                 }
+
+                // increment reading position
+                readingPosition++;
+            }
+
+            // Are there any bytes to move into the buffer?
+            if (readingPosition < bytesToRead) {
+                this.buffer = new Buffer(bytesToRead - readingPosition);
```

Comment:
With comments? `Buffer` is a [standard object provided by Node.js](https://nodejs.org/api/buffer.html) so we are creating a new `Buffer` is the size of how many bytes are left in the data buffer so we can move and store that into this.buffer for the next time this function is ran.